pkg/controller/direct/alloydb/instance_controller.go

```diff
@@ -0,0 +1,440 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alloydb
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/alloydb/v1beta1"
+	refsv1beta1 "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+
+	gcp "cloud.google.com/go/alloydb/apiv1beta"
+	alloydbpb "cloud.google.com/go/alloydb/apiv1beta/alloydbpb"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func init() {
+	registry.RegisterModel(krm.AlloyDBInstanceGVK, NewInstanceModel)
+}
+
+func NewInstanceModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &instanceModel{config: *config}, nil
+}
+
+var _ directbase.Model = &instanceModel{}
+
+type instanceModel struct {
+	config config.ControllerConfig
+}
+
+func (m *instanceModel) client(ctx context.Context) (*gcp.AlloyDBAdminClient, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewAlloyDBAdminRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("error building AlloyDB client for Instance: %w", err)
+	}
+	return gcpClient, err
+}
+
+func resolveInstanceType(ctx context.Context, reader client.Reader, obj *krm.AlloyDBInstance, isDeletion bool) error {
+	if obj.Spec.InstanceType == nil && obj.Spec.InstanceTypeRef == nil {
```

Comment:
We should not support `spec.instanceType` field in this resource, but only require `spec.instanceTypeRef`. I know the resource is already in Beta, can we mark the `spec.instanceType` as "not recommended", and make a backward compatibility support?

Reply:
Added "Not recommended." in the field description.

Reply:
This function needs to be modified as well to reflect "not recommended" otherwise it will still confuse users.
When we say "not recommended", it means the `instanceType` and `instanceTypeRef` is no longer a OneOf, but  if `instanceTypeRef` is configured, it is always more valued (we may warn users that `instanceType` is not needed if it is also configured).  If only instanceType is configured, we should *backport* the `instanceTypeRef` value (the `external` field). This guanratees the backward compatibility happens inside a single function rather than split everywhere when read/write to the type and typeRef.